File: apps/rolando_discord/lib/rolando_discord/commands.ex
defmodule RolandoDiscord.Commands do
  alias Nostrum.Api.ApplicationCommand
  alias Nostrum.Constants.ApplicationCommandOptionType, as: Opt
  require Logger

  def commands do
    [
      %{
        name: "train",
        description: "Fetches all available messages in the server to be used as training data",
        dm_permission: false
      },
      %{name: "gif", description: "Returns a gif from the ones it knows"},
      %{name: "image", description: "Returns an image from the ones it knows"},
      %{name: "video", description: "Returns a video from the ones it knows"},
      %{
        name: "analytics",
        description: "Returns analytics about the bot in this server"
      },
      %{
        name: "togglepings",
        description: "Toggles wether pings are enabled or not"
      },
      %{
        name: "replyrate",
        description: "View or set the reply rate",
        options: [
          %{
            type: Opt.integer(),
            name: "rate",
            description: "the rate to set (leave empty to view)",
            required: false
          }
        ]
      },
      %{
        name: "reactionrate",
        description: "View or set the reaction rate",
        options: [
          %{
            type: Opt.integer(),
            name: "rate",
            description: "the rate to set (leave empty to view)",
            required: false
          }
        ]
      },
      %{
        name: "cohesion",
        description:
          "View or set the cohesion value; A higher value makes sentences more coherent",
        options: [
          %{
            type: Opt.integer(),
            name: "value",
            description: "the value to set, must be at least 2 (leave empty to view)",
            required: false,
            min_value: 2,
            max_value: 10
          }
        ]
      },
      %{
        name: "opinion",
        description: "Generates a random opinion based on the provided seed",
        options: [
          %{
            type: Opt.string(),
            name: "about",
            description: "The seed of the message",
            required: true
          }
        ]
      },
      %{
        name: "wipe",
        description: "Deletes the given argument `data` from the training data",
        options: [
          %{
            type: Opt.string(),
            name: "data",
            description: "The data to be deleted",
            required: true
          }
        ]
      },
      %{
        name: "channels",
        description: "View which channels can be accessed for training data"
      },
      %{
        name: "vc",
        description: "Manage VC features",
        dm_permission: false,
        options: [
          # Subcommand: Join
          %{
            type: Opt.sub_command(),
            name: "join",
            description: "Joins the voice channel you are currently in"
          },
          # Subcommand: Leave
          %{
            type: Opt.sub_command(),
            name: "leave",
            description: "Leaves the voice channel you are currently in"
          },
          # Subcommand: Speak
          %{
            type: Opt.sub_command(),
            name: "speak",
            description: "Speaks words of wisdom in a VC, and then leaves",
            options: [
              %{
                type: Opt.channel(),
                name: "channel",
                description:
                  "Optional: Specific channel to speak in (defaults to the one you are connected to)",
                required: false
              }
            ]
          },
          # Subcommand: Joinrate View/Set
          %{
            type: Opt.sub_command(),
            name: "joinrate",
            description: "View or set the VC random join rate",
            options: [
              %{
                type: Opt.number(),
                name: "rate",
                description: "The rate in % (0-100)",
                required: false,
                min_value: 0.0,
                max_value: 100.0
              },
              %{
                type: Opt.channel(),
                name: "channel",
                description: "Optional: Specific channel to view / update (defaults to current)",
                required: false
              }
            ]
          },
          # Subcommand: Language View/Set
          %{
            type: Opt.sub_command(),
            name: "language",
            description: "View or set the VC language",
            options: [
              %{
                type: Opt.integer(),
                name: "language",
                description: "the language to set (leave empty to view)",
                required: false,
                choices: [
                  %{name: "English", value: 0},
                  %{name: "Italian", value: 1},
                  %{name: "German", value: 2},
                  %{name: "Spanish", value: 3}
                ]
              },
              %{
                type: Opt.channel(),
                name: "channel",
                description: "Optional: Specific channel to update (defaults to current)",
                required: false
              }
            ]
          }
        ]
      }
    ]
  end

  defp needs_resync?(remote, local) do
    cond do
      length(remote) != length(local) ->
        true

      true ->
        local_names = local |> Enum.map(& &1.name) |> Enum.sort()
        remote_names = remote |> Enum.map(& &1.name) |> Enum.sort()
        local_names != remote_names

        # TODO better diffing
    end
  end

  def sync do
    case ApplicationCommand.global_commands() do
      {:ok, remote_commands} ->
        if needs_resync?(remote_commands, commands()) do
          Logger.info("Syncing slash command interactions...")

          case sync_force() do
            {:ok, _} ->
              Logger.info("Commands synced successfully.")

            {:error, reason} ->
              Logger.error("Failed to sync commands: #{inspect(reason)}")
          end
        else
          Logger.info("Slash commands are up to date. Skipping sync.")
        end

      {:error, _} ->
        Logger.error("Could not verify commands")
    end
  end

  def sync_force do
    ApplicationCommand.bulk_overwrite_global_commands(commands())
  end
end

File: cmd/idiscord/commands/commands.go
package commands

import (
	"rolando/cmd/idiscord/services"
	"rolando/internal/config"
	"rolando/internal/logger"
	"time"

	"github.com/disgoorg/disgo/bot"
	"github.com/disgoorg/disgo/discord"
	"github.com/disgoorg/disgo/events"
)

type SlashCommandsHandler struct {
	Client        *bot.Client
	ChainsService *services.ChainsService
	Commands      map[string]SlashCommandHandler
}

type SlashCommandHandler func(client *bot.Client, event *events.ApplicationCommandInteractionCreate)

type SlashCommand struct {
	Command  discord.ApplicationCommandCreate
	Handler  SlashCommandHandler
	GuildIds []string // Optional guild IDs to restrict the command to specific guilds
}

// Constructor function for SlashCommandsHandler
func NewSlashCommandsHandler(
	client *bot.Client,
	chainsService *services.ChainsService,
) *SlashCommandsHandler {
	handler := &SlashCommandsHandler{
		Client:        client,
		ChainsService: chainsService,
		Commands:      make(map[string]SlashCommandHandler),
	}

	// Initialize commands
	handler.registerCommands([]SlashCommand{
		{
			Command: discord.SlashCommandCreate{
				Name:        "train",
				Description: "Fetches all available messages in the server to be used as training data",
				Contexts: []discord.InteractionContextType{
					discord.InteractionContextTypeGuild,
				},
			},
			Handler: handler.withAdminPermission(handler.trainCommand),
		},
		{
			Command: discord.SlashCommandCreate{
				Name:        "gif",
				Description: "Returns a gif from the ones it knows",
				Contexts: []discord.InteractionContextType{
					discord.InteractionContextTypeGuild,
				},
			},
			Handler: handler.gifCommand,
		},
		{
			Command: discord.SlashCommandCreate{
				Name:        "image",
				Description: "Returns an image from the ones it knows",
				Contexts: []discord.InteractionContextType{
					discord.InteractionContextTypeGuild,
				},
			},
			Handler: handler.imageCommand,
		},
		{
			Command: discord.SlashCommandCreate{
				Name:        "video",
				Description: "Returns a video from the ones it knows",
				Contexts: []discord.InteractionContextType{
					discord.InteractionContextTypeGuild,
				},
			},
			Handler: handler.videoCommand,
		},
		{
			Command: discord.SlashCommandCreate{
				Name:        "analytics",
				Description: "Returns analytics about the bot in this server",
				Contexts: []discord.InteractionContextType{
					discord.InteractionContextTypeGuild,
				},
			},
			Handler: handler.analyticsCommand,
		},
		{
			Command: discord.SlashCommandCreate{
				Name:        "togglepings",
				Description: "Toggles wether pings are enabled or not",
				Contexts: []discord.InteractionContextType{
					discord.InteractionContextTypeGuild,
				},
			},
			Handler: handler.withAdminPermission(handler.togglePingsCommand),
		},
		{
			Command: discord.SlashCommandCreate{
				Name:        "replyrate",
				Description: "View or set the reply rate",
				Contexts: []discord.InteractionContextType{
					discord.InteractionContextTypeGuild,
				},
				Options: []discord.ApplicationCommandOption{
					discord.ApplicationCommandOptionInt{
						Name:        "rate",
						Description: "the rate to set (leave empty to view)",
						Required:    false,
					},
				},
			},
			Handler: handler.replyRateCommand,
		},
		{
			Command: discord.SlashCommandCreate{
				Name:        "reactionrate",
				Description: "View or set the reaction rate",
				Contexts: []discord.InteractionContextType{
					discord.InteractionContextTypeGuild,
				},
				Options: []discord.ApplicationCommandOption{
					discord.ApplicationCommandOptionInt{
						Name:        "rate",
						Description: "the rate to set (leave empty to view)",
						Required:    false,
					},
				},
			},
			Handler: handler.reactionRateCommand,
		},
		{
			Command: discord.SlashCommandCreate{
				Name:        "cohesion",
				Description: "View or set the cohesion value; A higher value makes sentences more coherent",
				Contexts: []discord.InteractionContextType{
					discord.InteractionContextTypeGuild,
				},
				Options: []discord.ApplicationCommandOption{
					discord.ApplicationCommandOptionInt{
						MinValue:    new(2),
						MaxValue:    new(255),
						Name:        "value",
						Description: "the value to set, must be at least 2 (leave empty to view)",
						Required:    false,
					},
				},
			},
			Handler: handler.cohesionCommand,
		},
		{
			Command: discord.SlashCommandCreate{
				Name:        "opinion",
				Description: "Generates a random opinion based on the provided seed",
				Contexts: []discord.InteractionContextType{
					discord.InteractionContextTypeGuild,
				},
				Options: []discord.ApplicationCommandOption{
					discord.ApplicationCommandOptionString{
						Name:        "about",
						Description: "The seed of the message",
						Required:    true,
					},
				},
			},
			Handler: handler.opinionCommand,
		},
		{
			Command: discord.SlashCommandCreate{
				Name:        "wipe",
				Description: "Deletes the given argument `data` from the training data",
				Contexts: []discord.InteractionContextType{
					discord.InteractionContextTypeGuild,
				},
				Options: []discord.ApplicationCommandOption{
					discord.ApplicationCommandOptionString{
						Name:        "data",
						Description: "The data to be deleted",
						Required:    true,
					},
				},
			},
			Handler: handler.wipeCommand,
		},
		{
			Command: discord.SlashCommandCreate{
				Name:        "channels",
				Description: "View which channels can be accessed for training data",
				Contexts: []discord.InteractionContextType{
					discord.InteractionContextTypeGuild,
				},
			},
			Handler: handler.channelsCommand,
		},
		{
			Command: discord.SlashCommandCreate{
				Name:        "src",
				Description: "Provides the URL to the repository with bot source code",
				Contexts: []discord.InteractionContextType{
					discord.InteractionContextTypeGuild,
				},
			},
			Handler: handler.srcCommand,
		},
		{
			Command: discord.SlashCommandCreate{
				Name:        "vc-join",
				Description: "Joins the voice channel you are currently in",
				Contexts: []discord.InteractionContextType{
					discord.InteractionContextTypeGuild,
				},
			},
			Handler: handler.withGuildSubscription(config.VoiceChatFeaturesSKU, handler.vcJoinCommand),
		},
		{
			Command: discord.SlashCommandCreate{
				Name:        "vc-leave",
				Description: "Leaves the voice channel you are currently in",
				Contexts: []discord.InteractionContextType{
					discord.InteractionContextTypeGuild,
				},
			},
			Handler: handler.withGuildSubscription(config.VoiceChatFeaturesSKU, handler.vcLeaveCommand),
		},
		{
			Command: discord.SlashCommandCreate{
				Name:        "vc-speak",
				Description: "Speaks a message in the VC you are in, and then leaves",
				Contexts: []discord.InteractionContextType{
					discord.InteractionContextTypeGuild,
				},
			},
			Handler: handler.withGuildSubscription(config.VoiceChatFeaturesSKU, handler.vcSpeakCommand),
		},
		{
			Command: discord.SlashCommandCreate{
				Name:        "vc-language",
				Description: "View or set the language to use when generating audio",
				Contexts: []discord.InteractionContextType{
					discord.InteractionContextTypeGuild,
				},
				Options: []discord.ApplicationCommandOption{
					discord.ApplicationCommandOptionInt{
						Name: "language",
						Choices: []discord.ApplicationCommandOptionChoiceInt{
							{
								Name:  "English",
								Value: 0,
							},
							{
								Name:  "Italian",
								Value: 1,
							},
							{
								Name:  "German",
								Value: 2,
							},
							{
								Name:  "Spanish",
								Value: 3,
							},
						},
						Description: "the language to set (leave empty to view)",
						Required:    false,
					},
				},
			},
			Handler: handler.withGuildSubscription(config.VoiceChatFeaturesSKU, handler.vcLanguageCommand),
		},
		{
			Command: discord.SlashCommandCreate{
				Name:        "vc-joinrate",
				Description: "View or set the VC random join rate",
				Contexts: []discord.InteractionContextType{
					discord.InteractionContextTypeGuild,
				},
				Options: []discord.ApplicationCommandOption{
					discord.ApplicationCommandOptionInt{
						Name:        "rate",
						Description: "the rate to set (1/rate) (leave empty to view)",

						Required: false,
					},
				},
			},
			Handler: handler.withGuildSubscription(config.VoiceChatFeaturesSKU, handler.vcJoinRateCommand),
		},
	})

	return handler
}

func commandCreates(cmds []SlashCommand) []discord.ApplicationCommandCreate {
	out := make([]discord.ApplicationCommandCreate, len(cmds))
	for i := range cmds {
		out[i] = cmds[i].Command
	}
	return out
}

func (h *SlashCommandsHandler) registerCommands(commands []SlashCommand) {
	for i := range commands {
		h.Commands[commands[i].Command.CommandName()] = commands[i].Handler
	}

	var global []SlashCommand
	guildSeen := make(map[string]struct{})
	for i := range commands {
		if len(commands[i].GuildIds) == 0 {
			global = append(global, commands[i])
			continue
		}
		for _, gid := range commands[i].GuildIds {
			guildSeen[gid] = struct{}{}
		}
	}

	if len(global) <= 0 {
		logger.Infof("No global slash commands to sync.")
		return
	}

	refresh := config.ForceCommandRefresh
	if refresh {
		logger.Infof("Syncing global slash command interactions...")
		_, err := h.Client.Rest.SetGlobalCommands(h.Client.ApplicationID, commandCreates(global))
		if err != nil {
			logger.Errorf("Failed to sync global slash commands: %v", err)
		} else {
			logger.Infof("Global slash commands synced successfully.")
		}
	}

	// for gid := range guildSeen {
	// 	var guildCmds []SlashCommand
	// 	for i := range commands {
	// 		if slices.Contains(commands[i].GuildIds, gid) {
	// 			guildCmds = append(guildCmds, commands[i])
	// 		}
	// 	}
	// 	guildID := snowflake.MustParse(gid)
	// 	remote, err := h.Client.Rest.GetGuildCommands(h.Client.ApplicationID, guildID, false)
	// 	if err != nil {
	// 		logger.Errorf("Could not verify guild %s slash commands: %v", gid, err)
	// 		continue
	// 	}
	// 	if force || needsResync(remote, guildCmds) {
	// 		logger.Infof("Syncing guild %s slash command interactions...", gid)
	// 		_, err := h.Client.Rest.SetGuildCommands(h.Client.ApplicationID, guildID, commandCreates(guildCmds))
	// 		if err != nil {
	// 			logger.Errorf("Failed to sync guild %s slash commands: %v", gid, err)
	// 		} else {
	// 			logger.Infof("Guild %s slash commands synced successfully.", gid)
	// 		}
	// 	} else {
	// 		logger.Infof("Guild %s slash commands are up to date. Skipping sync.", gid)
	// 	}
	// }
}

// Entry point for handling slash command interactions
func (h *SlashCommandsHandler) OnSlashCommandInteraction(event *events.ApplicationCommandInteractionCreate) {
	commandName := event.SlashCommandInteractionData().CommandName()

	var where string
	if event.GuildID() == nil {
		where = "DMs"
	} else {
		if guild, ok := h.Client.Caches.Guild(*event.GuildID()); ok {
			where = guild.Name
		} else {
			logger.Errorf("Failed to fetch guild '%s' for command interaction", event.GuildID().String())
			return
		}
	}

	who := event.User().EffectiveName()

	startTime := time.Now()
	logger.Infof("from '%s' in '%s': /%s", who, where, commandName)
	if handler, exists := h.Commands[commandName]; exists {
		handler(h.Client, event) // Call the function bound to this command
	}
	logger.Infof("/%s handler from '%s' in '%s' done in %s", commandName, who, where, time.Since(startTime).String())
}
